logo.js

```diff
@@ -841,19 +911,78 @@ function LogoInterpreter(turtle, stream, savehook)
       var atoms = parse(string);
 
       // And execute it!
-      return self.execute(atoms, options);
+      var promise = self.execute(atoms, options).then(null, function (err) {
+        if (err && err.special == "bye") {
+          return undefined;
+        }
+        if (self.turtle) {
+          self.turtle.end();
+        }
+        throw err;
+      });
+      this._setPendingPromise(promise);
+      self._pendingPromise = promise;
+      return promise;
     } catch (e) {
+      // FIXME: doesn't work with promises
       if (e instanceof Bye) {
         // clean exit
-        return undefined;
+        return Promise.resolve(undefined);
       } else {
-        throw e;
+        return Promise.reject(e);
       }
     } finally {
       if (self.turtle) { self.turtle.end(); }
     }
   };
 
+  //----------------------------------------------------------------------
+  // Helpers to ensure calls to run() are serialized:
+  //----------------------------------------------------------------------
+
+  // When there's an outstanding call to .run() (whose promise hasn't resolved
```

Comment:
We could make the assertions implicitly chain within a test - e.g. each one pushes onto a queue which is run sequentially when the test body returns.